File: bindings/elixir_project/lib/hako_binary/binary_io.ex
defmodule PduConversion do
  def bin_to_int8(<<value::signed-8>>), do: value
  def bin_to_int16(<<value::signed-16-little>>), do: value
  def bin_to_int32(<<value::signed-32-little>>), do: value
  def bin_to_bool(<<value::signed-32-little>>), do: value
  def bin_to_int64(<<value::signed-64-little>>), do: value
  def bin_to_uint8(<<value::unsigned-8>>), do: value
  def bin_to_uint16(<<value::unsigned-16-little>>), do: value
  def bin_to_uint32(<<value::unsigned-32-little>>), do: value
  def bin_to_uint64(<<value::unsigned-64-little>>), do: value
  def bin_to_float32(<<value::float-32-little>>), do: value
  def bin_to_float64(<<value::float-64-little>>), do: value

  def bin_to_string(binary), do: binary |> :binary.bin_to_list() |> Enum.filter(&(&1 != 0)) |> :binary.list_to_bin()

  def int8_to_bin(value), do: <<value::signed-8>>
  def int16_to_bin(value), do: <<value::signed-16-little>>
  def int32_to_bin(value), do: <<value::signed-32-little>>
  def bool_to_bin(value), do: <<value::signed-32-little>>
  def int64_to_bin(value), do: <<value::signed-64-little>>
  def uint8_to_bin(value), do: <<value::unsigned-8>>
  def uint16_to_bin(value), do: <<value::unsigned-16-little>>
  def uint32_to_bin(value), do: <<value::unsigned-32-little>>
  def uint64_to_bin(value), do: <<value::unsigned-64-little>>
  def float32_to_bin(value), do: <<value::float-32-little>>
  def float64_to_bin(value), do: <<value::float-64-little>>
  def string_to_bin(value), do: :binary.list_to_bin(:binary.bin_to_list(value) ++ [0])

  def type_to_bin("int8", value), do: int8_to_bin(value)
  def type_to_bin("uint8", value), do: uint8_to_bin(value)
  def type_to_bin("int16", value), do: int16_to_bin(value)
  def type_to_bin("uint16", value), do: uint16_to_bin(value)
  def type_to_bin("int32", value), do: int32_to_bin(value)
  def type_to_bin("bool", value), do: bool_to_bin(value)
  def type_to_bin("uint32", value), do: uint32_to_bin(value)
  def type_to_bin("int64", value), do: int64_to_bin(value)
  def type_to_bin("uint64", value), do: uint64_to_bin(value)
  def type_to_bin("float32", value), do: float32_to_bin(value)
  def type_to_bin("float64", value), do: float64_to_bin(value)
  def type_to_bin("string", value), do: string_to_bin(value)
  def type_to_bin(_type, _value), do: nil

  def bin_to_value("int8", value), do: bin_to_int8(value)
  def bin_to_value("uint8", value), do: bin_to_uint8(value)
  def bin_to_value("int16", value), do: bin_to_int16(value)
  def bin_to_value("uint16", value), do: bin_to_uint16(value)
  def bin_to_value("int32", value), do: bin_to_int32(value)
  def bin_to_value("bool", value), do: bin_to_bool(value)
  def bin_to_value("uint32", value), do: bin_to_uint32(value)
  def bin_to_value("int64", value), do: bin_to_int64(value)
  def bin_to_value("uint64", value), do: bin_to_uint64(value)
  def bin_to_value("float32", value), do: bin_to_float32(value)
  def bin_to_value("float64", value), do: bin_to_float64(value)
  def bin_to_value("string", value), do: bin_to_string(value)
  def bin_to_value(_type, _value), do: nil

  def bin_to_array_values("int8", value), do: for(<<x::signed-8 <- value>>, do: x)
  def bin_to_array_values("uint8", value), do: for(<<x::unsigned-8 <- value>>, do: x)
  def bin_to_array_values("int16", value), do: for(<<x::signed-16-little <- value>>, do: x)
  def bin_to_array_values("uint16", value), do: for(<<x::unsigned-16-little <- value>>, do: x)
  def bin_to_array_values("int32", value), do: for(<<x::signed-32-little <- value>>, do: x)
  def bin_to_array_values("bool", value), do: for(<<x::signed-32-little <- value>>, do: x)
  def bin_to_array_values("uint32", value), do: for(<<x::unsigned-32-little <- value>>, do: x)
  def bin_to_array_values("int64", value), do: for(<<x::signed-64-little <- value>>, do: x)
  def bin_to_array_values("uint64", value), do: for(<<x::unsigned-64-little <- value>>, do: x)
  def bin_to_array_values("float32", value), do: for(<<x::float-32-little <- value>>, do: x)
  def bin_to_array_values("float64", value), do: for(<<x::float-64-little <- value>>, do: x)

  def bin_to_array_values(_type, _value), do: nil

  def write_binary(binary_data, off, bin) do
    Enum.with_index(bin)
    |> Enum.reduce(binary_data, fn {data, i}, acc ->
      <<before::binary-size(off + i), _::binary>> = acc
      <<before::binary, data::binary, acc::binary>>
    end)
  end

  def read_binary(binary_data, off, size) do
    binary_part(binary_data, off, size)
  end
end
defmodule PduMetaData do
  @pdu_meta_data_size 24
  @pdu_meta_data_magicno 0x12345678
  @pdu_meta_data_version 1

  defstruct [:magicno, :version, :base_off, :heap_off, :total_size]

  def new do
    %PduMetaData{
      magicno: 0,
      version: 0,
      base_off: 0,
      heap_off: 0,
      total_size: 0
    }
  end

  def set_empty(%PduMetaData{} = meta) do
    %PduMetaData{
      meta
      | magicno: @pdu_meta_data_magicno,
        version: @pdu_meta_data_version,
        base_off: @pdu_meta_data_size
    }
  end

  def to_bytes(%PduMetaData{} = meta) do
    <<meta.magicno::32-little, meta.version::32-little, meta.base_off::32-little,
      meta.heap_off::32-little, meta.total_size::32-little>>
  end

  def get_pdu_meta_data_size, do: @pdu_meta_data_size
  def get_pdu_meta_data_magicno, do: @pdu_meta_data_magicno
  def get_pdu_meta_data_version, do: @pdu_meta_data_version
end


defmodule PduMetaDataParser do
  def load_pdu_meta(binary_data) do
    if byte_size(binary_data) < PduMetaData.get_pdu_meta_data_size() do
      {:error, :invalid_data}
    else
      <<magicno::32-little, version::32-little, base_off::32-little,
        heap_off::32-little, total_size::32-little>> = binary_data

      if magicno != PduMetaData.get_pdu_meta_data_magicno() or version != PduMetaData.get_pdu_meta_data_version() do
        {:error, :invalid_magicno_or_version}
      else
        {:ok,
         %PduMetaData{
           magicno: magicno,
           version: version,
           base_off: base_off,
           heap_off: heap_off,
           total_size: total_size
         }}
      end
    end
  end
end

File: bindings/python/hako_binary/binary_io.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import glob
import re
import os
from sys import byteorder
import struct

def binToint8(binary):
    return struct.unpack('b', binary)[0]

def binToint16(binary):
    return struct.unpack('h', binary)[0]

def binToint32(binary):
    return struct.unpack('i', binary)[0]

def binTobool(binary):
    return struct.unpack('i', binary)[0]

def binToint64(binary):
    return struct.unpack('q', binary)[0]

def binTouint8(binary):
    return struct.unpack('B', binary)[0]

def binTouint16(binary):
    return struct.unpack('H', binary)[0]

def binTouint32(binary):
    return struct.unpack('I', binary)[0]

def binTouint64(binary):
    return struct.unpack('Q', binary)[0]

def binTofloat32(binary):
    return struct.unpack('f', binary)[0]

def binTofloat64(binary):
    return struct.unpack('d', binary)[0]

def binTostring(binary, max_len=128):
    try:
        sub = binary[:max_len]
        end = sub.find(b'\0')
        if end == -1:
            end = max_len
        return sub[:end].decode('utf-8')
    except UnicodeDecodeError as e:
        print("UnicodeDecodeError:")
        print(f"  ERROR: {e}")
        print(f"  RAW DATA: {binary}")
        print(f"  HEXDUMP: {' '.join(f'{b:02x}' for b in binary)}")
        raise e



def int8Tobin(arg):
    return struct.pack('b', arg)

def int16Tobin(arg):
    return struct.pack('h', arg)

def int32Tobin(arg):
    return struct.pack('i', arg)

def boolTobin(arg):
    return struct.pack('i', arg)

def int64Tobin(arg):
    return struct.pack('q', arg)

def uint8Tobin(arg):
    return struct.pack('B', arg)

def uint16Tobin(arg):
    return struct.pack('H', arg)

def uint32Tobin(arg):
    return struct.pack('I', arg)

def uint64Tobin(arg):
    return struct.pack('Q', arg)

def float32Tobin(arg):
    return struct.pack('f', arg)

def float64Tobin(arg):
    return struct.pack('d', arg)

def stringTobin(arg):
    return arg.encode(encoding='utf-8') + b'\x00'

def typeTobin(type, arg):
    if (type == "int8"):
        return int8Tobin(arg)
    elif (type == "uint8"):
        return uint8Tobin(arg)
    elif (type == "int16"):
        return int16Tobin(arg)
    elif (type == "uint16"):
        return uint16Tobin(arg)
    elif (type == "int32"):
        return int32Tobin(arg)
    elif (type == "bool"):
        return boolTobin(arg)
    elif (type == "uint32"):
        return uint32Tobin(arg)
    elif (type == "int64"):
        return int64Tobin(arg)
    elif (type == "uint64"):
        return uint64Tobin(arg)
    elif (type == "float32"):
        return float32Tobin(arg)
    elif (type == "float64"):
        return float64Tobin(arg)
    elif (type == "string"):
        return stringTobin(arg)
    else:
        return None

def binTovalue(type, arg):
    if (type == "int8"):
        return binToint8(arg)
    elif (type == "uint8"):
        return binTouint8(arg)
    elif (type == "int16"):
        return binToint16(arg)
    elif (type == "uint16"):
        return binTouint16(arg)
    elif (type == "int32"):
        return binToint32(arg)
    elif (type == "bool"):
        return binTobool(arg)
    elif (type == "uint32"):
        return binTouint32(arg)
    elif (type == "int64"):
        return binToint64(arg)
    elif (type == "uint64"):
        return binTouint64(arg)
    elif (type == "float32"):
        return binTofloat32(arg)
    elif (type == "float64"):
        return binTofloat64(arg)
    elif (type == "string"):
        return binTostring(arg)
    else:
        return None
def binToArrayValues(type, arg):
    # little endian
    if (type == "int8"):
        return struct.unpack(f'<{len(arg)}b', arg)
    elif (type == "uint8"):
        return struct.unpack(f'<{len(arg)}B', arg)
    elif (type == "int16"):
        return struct.unpack(f'<{len(arg)//2}h', arg)
    elif (type == "uint16"):
        return struct.unpack(f'<{len(arg)//2}H', arg)
    elif (type == "int32"):
        return struct.unpack(f'<{len(arg)//4}i', arg)
    elif (type == "bool"):
        return struct.unpack(f'<{len(arg)//4}i', arg)
    elif (type == "uint32"):
        return struct.unpack(f'<{len(arg)//4}I', arg)
    elif (type == "int64"):
        return struct.unpack(f'<{len(arg)//8}q', arg)
    elif (type == "uint64"):
        return struct.unpack(f'<{len(arg)//8}Q', arg)
    elif (type == "float32"):
        return struct.unpack(f'<{len(arg)//4}f', arg)
    elif (type == "float64"):
        return struct.unpack(f'<{len(arg)//8}d', arg)
    else:
        return None

def typeTobin_array(type, values, elm_size=None):
    count = len(values)
    if type == "int8":
        return struct.pack(f'<{count}b', *values)
    elif type == "uint8":
        return struct.pack(f'<{count}B', *values)
    elif type == "int16":
        return struct.pack(f'<{count}h', *values)
    elif type == "uint16":
        return struct.pack(f'<{count}H', *values)
    elif type == "int32":
        return struct.pack(f'<{count}i', *values)
    elif type == "bool":
        return struct.pack(f'<{count}i', *values)
    elif type == "uint32":
        return struct.pack(f'<{count}I', *values)
    elif type == "int64":
        return struct.pack(f'<{count}q', *values)
    elif type == "uint64":
        return struct.pack(f'<{count}Q', *values)
    elif type == "float32":
        return struct.pack(f'<{count}f', *values)
    elif type == "float64":
        return struct.pack(f'<{count}d', *values)
    elif type == "string":
        if elm_size is None:
            raise ValueError("elm_size required for string array")
        binaries = []
        for val in values:
            raw = val.encode('utf-8') + b'\x00'
            if len(raw) > elm_size:
                raw = raw[:elm_size]
            buffer = bytearray(elm_size)
            buffer[:len(raw)] = raw
            binaries.append(buffer)
        return b''.join(binaries)
    else:
        raise ValueError(f"typeTobin_array: Unsupported type {type}")


def writeBinary(binary_data, off, bin):
    i = 0
    for data in bin:
        binary_data[off + i] = data
        i = i + 1

def readBinary(binary_data, off, size):
    return binary_data[off:off+size]

class PduMetaData:
    PDU_META_DATA_SIZE = 24
    PDU_META_DATA_MAGICNO = 0x12345678
    PDU_META_DATA_VERSION = 1
    def __init__(self):
        self.magicno = 0
        self.version = 0
        self.base_off = 0
        self.heap_off = 0
        self.total_size = 0
    def set_empty(self):
        self.magicno = PduMetaData.PDU_META_DATA_MAGICNO
        self.version = PduMetaData.PDU_META_DATA_VERSION
        self.base_off = PduMetaData.PDU_META_DATA_SIZE
    def to_bytes(self):
        data = bytearray()
        data.extend(self.magicno.to_bytes(4, byteorder='little'))
        data.extend(self.version.to_bytes(4, byteorder='little'))
        data.extend(self.base_off.to_bytes(4, byteorder='little'))
        data.extend(self.heap_off.to_bytes(4, byteorder='little'))
        data.extend(self.total_size.to_bytes(4, byteorder='little'))
        return data


class PduMetaDataParser:
    def __init__(self):
        self.meta = PduMetaData()

    def load_pdu_meta(self, binary_data):
        if len(binary_data) < PduMetaData.PDU_META_DATA_SIZE:
            return None
        magicno = binTovalue("uint32", readBinary(binary_data, 0, 4))
        version = binTovalue("uint32", readBinary(binary_data, 4, 4))
        if magicno != PduMetaData.PDU_META_DATA_MAGICNO or version != PduMetaData.PDU_META_DATA_VERSION:
            return None
        self.meta.magicno    = magicno
        self.meta.version    = version
        self.meta.base_off   = binTovalue("uint32", readBinary(binary_data, 8, 4))
        self.meta.heap_off   = binTovalue("uint32", readBinary(binary_data, 12, 4))
        self.meta.total_size = binTovalue("uint32", readBinary(binary_data, 16, 4))
        return self.meta
